// @theia/debug/lib/browser/view/debug-exception-breakpoint.js
"use strict";
// *****************************************************************************
// Copyright (C) 2019 TypeFox and others.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License v. 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0.
//
// This Source Code may also be made available under the following Secondary
// Licenses when the conditions for such availability set forth in the Eclipse
// Public License v. 2.0 are satisfied: GNU General Public License, version 2
// with the GNU Classpath Exception which is available at
// https://www.gnu.org/software/classpath/license.html.
//
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
// *****************************************************************************
Object.defineProperty(exports, "__esModule", { value: true });
exports.DebugExceptionBreakpoint = void 0;
const React = require("@theia/core/shared/react");
const browser_1 = require("@theia/core/lib/browser");
const tree_selection_1 = require("@theia/core/lib/browser/tree/tree-selection");
const core_1 = require("@theia/core");
const debug_commands_1 = require("../debug-commands");
class DebugExceptionBreakpoint {
    constructor(data, breakpoints, commandService) {
        this.data = data;
        this.breakpoints = breakpoints;
        this.commandService = commandService;
        this.onEdit = async () => {
            await this.selectInTree();
            this.commandService.executeCommand(debug_commands_1.DebugCommands.EDIT_BREAKPOINT_CONDITION.id);
        };
        this.toggle = () => this.breakpoints.toggleExceptionBreakpoint(this.data.raw.filter);
        this.id = data.raw.filter + ':' + data.raw.label;
    }
    render(host) {
        this.treeWidget = host;
        return React.createElement("div", { title: this.data.raw.description || this.data.raw.label, className: 'theia-source-breakpoint' },
            React.createElement("span", { className: 'theia-debug-breakpoint-icon' }),
            React.createElement("input", { type: 'checkbox', checked: this.data.enabled, onChange: this.toggle }),
            React.createElement("span", { className: 'line-info' },
                React.createElement("span", { className: 'name' },
                    this.data.raw.label,
                    " "),
                this.data.condition &&
                    React.createElement("span", { title: core_1.nls.localizeByDefault('Expression condition: {0}', this.data.condition), className: 'path ' + browser_1.TREE_NODE_INFO_CLASS },
                        this.data.condition,
                        " ")),
            this.renderActions());
    }
    renderActions() {
        if (this.data.raw.supportsCondition) {
            return React.createElement("div", { className: 'theia-debug-breakpoint-actions' },
                React.createElement("div", { className: (0, browser_1.codicon)('edit', true), title: core_1.nls.localizeByDefault('Edit Condition...'), onClick: this.onEdit }));
        }
        return undefined;
    }
    async selectInTree() {
        var _a;
        if (((_a = this.treeWidget) === null || _a === void 0 ? void 0 : _a.model) && tree_selection_1.SelectableTreeNode.is(this)) {
            this.treeWidget.model.selectNode(this);
        }
    }
    async editCondition() {
        const inputDialog = new browser_1.SingleTextInputDialog({
            title: this.data.raw.label,
            placeholder: this.data.raw.conditionDescription,
            initialValue: this.data.condition
        });
        let condition = await inputDialog.open();
        if (condition === undefined) {
            return;
        }
        if (condition === '') {
            condition = undefined;
        }
        if (condition !== this.data.condition) {
            this.breakpoints.updateExceptionBreakpoint(this.data.raw.filter, { condition });
        }
    }
}
exports.DebugExceptionBreakpoint = DebugExceptionBreakpoint;
